### DISCUSSION TITLE
Modular increment/decrement operators

### DISCUSSION THREAD

**Discussion by Rabadash8820** [dotnet/csharplang#8133] [Language Ideas]
I often find myself writing code that has to traverse a list and "loop around". E.g., in game code, I might want to pick 5 random entities from a list of N to initiate some behavior; a naive/performant implementation is to pick 5 elements starting from a random index in the list, with modulus to avoid `IndexOutOfBoundsException`. Having a single operator to do this "increment/decrement index but loop around" would be quite handy.

Here is the idea for the "modular increment operator":

```cs
int index = <random #>;

// Current C# operators:
index = (index + 1) % count;
// or
index++;
index %= count;

// Suggested operator
index ++%= count;
```

A "modular decrement operator", `--%=` would work similarly. Its implementation may not be as simple as `index = (index - 1) % count` as it would have to account for negative numbers (I typically use `index = (index - 1 + count) % count` in these situations). I think it could be implemented as `index = (index - 1) %% count` if I understand the proposal in #4744 correctly.

I admit, a 4-character operator is kinda ridiculous, and incrementing/decrementing by a number other than 1 would still require the same code as before. I did consider a ternary assignment operator like `index +%= delta count` but I'm not sure that's any more readable than current C# options. Anyway, just wanted to see if anyone else has considered this idea or would find it worthwhile.

**Comment by CyrusNajmabadi** [comment]
could do something like:

```c#
    public static void IncrementAndWrap(this ref int index, int count)
    {
        index = (index + 1) % count;
    }
```

**Reply by Rabadash8820** [reply]
True, thanks. Always cool to see a `ref` use case. It would be nice to avoid the method-call overhead for something this small, but some `MathUtil` static class could have a method like this. I'd probably just stick with inlining over this though. 🤷‍♂️

**Reply by AartBluestoke** [reply]
probably no overhead at all- the compiler trivially in-lines this.

**Reply by colejohnson66** [reply]
Indeed, [trivially inlined](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACY8pJ0hgYQBs9caBvGg2FNGzVgEsAdhk4R8AB2xQJuCFIAU02QgCUQkYOoiTTAOwMNCBgGoG5XQwCkDFAG4DwgL6eGv+kwsDNoMACowuBhaMgx6vkamIggAdACSUmCw+DAyAIJSACYA6lDYChoouh7Gieax1SY+NcL+YkHEKAzpmTDZeYUlZRoYABaqDLAAZsEx0gUwCGgzspAArjL6zQwJtXMLDAC8lnvWdg7ODGsyDSJNXkA), even at tier 0. Don't take the JIT (or NativeAOT) for granted; It's as powerful as your normal C compiler when it comes to analysis.

**Reply by Rabadash8820** [reply]
Wo, never seen sharplab before. Thanks for the link @colejohnson66, and for verifying this!
